Fix derive_next_unused

rajarshimaitra opened the pull request:
Fixes #110 

I think there was an internal logic problem previously where we were calling `next` twice. Once to check if next is available, and then in the return call in `else` block. This would progress the iterator twice on each call..

Restructured the logic to only call `next` once, and also solve the lifetime problem. This also allows us to remove redundant `clone`s elsewhere.

Renamed `derive_next_unused` to `next_unused`.

evanlinjin commented:
> I think there was an internal logic problem previously where we were calling next twice. Once to check if next is available, and then in the return call in else block. This would progress the iterator twice on each call..

The `keychain_unused` method is non-mutating, and the iterator returned on each call is different.

> Restructured the logic to only call next once, and also solve the lifetime problem. This also allows us to remove redundant clones elsewhere.

I like this. I would prefer if `derive_new` uses the old `new_spk` (instead of calling `self.inner.script_pubkeys().get(...)`.

Could you rebase your work on top of #116 ? Maybe create a PR onto my fork, thank you.

evanlinjin commented on bdk_chain/src/keychain/keychain_txout_index.rs:
```diff
@@ -237,21 +238,24 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
+            self.keychain_unused(keychain)
+                .next()
+                .map(|(i, s)| (i, s.to_owned()))
+        };
```
NIT: We don't need these extra brackets.

rajarshimaitra replied:
I didn't get it.. Which extra brackets?

evanlinjin commented on bdk_chain/src/keychain/keychain_txout_index.rs:
```diff
@@ -3,6 +3,7 @@ use crate::{
     miniscript::{Descriptor, DescriptorPublicKey},
     ForEachTxout, SpkTxOutIndex,
 };
+use alloc::borrow::ToOwned;
```
NIT: We don't need to use `ToOwned` as we are just going from `&T` to `T`.

rajarshimaitra replied:
Went ahead with a clone..

rajarshimaitra commented:
> I like this. I would prefer if derive_new uses the old new_spk (instead of calling self.inner.script_pubkeys().get(...).

Updated.. 

> Could you rebase your work on top of https://github.com/LLFourn/bdk_core_staging/pull/116 ? Maybe create a PR onto my fork, thank you.

Rebased on top of #116

evanlinjin commented on bdk_chain/src/keychain/keychain_txout_index.rs:
```diff
@@ -231,27 +231,25 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
-        if need_new {
-            self.derive_new(keychain)
+    pub fn next_unused<'a>(&'a mut self, keychain: &K) -> (u32, Script) {
+        let existing = {
+            self.keychain_unused(keychain)
+                .next()
+                .map(|(i, s)| (i, s.clone()))
+        };
```
> I didn't get it.. Which extra brackets?

@rajarshimaitra 

```suggestion
        let existing = self.keychain_unused(keychain)
            .next()
            .map(|(i, s)| (i, s.clone()));
```

rajarshimaitra replied:
Ah right.. I used that to solve that lifetime issue to remove the immutable ref first out of scope,  before taking another mutable ref.. But we don't that anymore as we are cloning data out..

rajarshimaitra replied:
Updated in https://github.com/LLFourn/bdk_core_staging/pull/117/commits/a47909ca82ca7b240b32cdfb9a391a5c3a83d722

LLFourn commented:
I think we should not clone the spks out here. It's easy for the caller to do this and it indicates that his was not just derived but also stored in the thing itself which is difficult to convey through the name of the method itself. I think maybe you can do it in a less awkward way without double inserting with the `Entry` API?

rajarshimaitra commented:
> I think we should not clone the spks out here.

The major hurdle I was facing for returning the reference is, that enforces the lifetime within the function scope to be specified such that two different references doesn't fall under same life time. This is a very edgy case of lifetimes and demonstrated in only documented this [rustnomicon post](https://doc.rust-lang.org/nomicon/lifetime-mismatch.html). We cannot take the 2nd mutable reference, without dropping the 1st immutable one, or some how define the lifetime in the scope such that 2 refs don't occur in the same lifetime.. Or else compiler won't allow passing out the ref..

> I think maybe you can do it in a less awkward way without double inserting with the Entry API?

That was my initial approach but it didn't seem to solve the above lifetime problem..

Thus so far "scoping out and copying data" seemed like the only non nasty fix to me.. But ya, this essentially doesn't allow us to pass out the ref any more.. The other option was what you had before,. have double calls and let go of the first call.

LLFourn commented:
> Another reason I went with copying was we were doing that in most of our call sites anyway, because that's generally is the downstream requirements for many script APIs.

I think the main API we care about here is the `Address` API which takes a reference. I'd be in favor of just doing the rename in this PR.

I'm not sure why @evanlinjin asked you to rebase on top of his draft PR? Can we just drop the changes other than the rename here and merge it.

rajarshimaitra commented:
@LLFourn reverted back to just renaming..

LLFourn commented on bdk_cli_lib/src/lib.rs:
```diff
@@ -319,9 +319,7 @@ pub fn create_tx<P: ChainPosition>(
-        let (index, script) = keychain_tracker
-            .txout_index
-            .derive_next_unused(&internal_keychain);
+        let (index, script) = keychain_tracker.txout_index.next_unused(&internal_keychain);
```
I just realized that we have to persist the keychain index increment here!!!!!
I actually noticed this bug we were missing a change output in my local esplora example db.

I'll try and fix this tomorrow.

LLFourn reviewed (approved):
Thanks @rajarshimaitra

evanlinjin commented:
> I'm not sure why @evanlinjin asked you to rebase on top of his draft PR? Can we just drop the changes other than the rename here and merge it.

The reason is that #116 already takes care of #110 so this is essentially a duplicate effort, except for the changes to reference/no-reference.

LLFourn commented:
Ok well since #116 has a different focus we'll just merge this now and move on.
